src/plugins/usage_collection/server/collector/collector_set.ts

```diff
@@ -92,51 +96,72 @@ export class CollectorSet {
+    if (!(collectors instanceof Map)) {
       throw new Error(
-        `areAllCollectorsReady method given bad collectorSet parameter: ` + typeof collectorSet
+        `areAllCollectorsReady method given bad Map of collectors: ` + typeof collectors
       );
     }
 
-    const collectors = [...collectorSet.collectors.values()];
-    const collectorsWithStatus = await Promise.all(
-      collectors.map(async (collector) => {
-        return {
-          isReady: await collector.isReady(),
-          collector,
-        };
+    const secondInMs = 1000;
+    const collectorsWithStatus: CollectorWithStatus[] = await Promise.all(
+      [...collectors.values()].map(async (collector) => {
+        const isReadyWithTimeout = await withTimeout<boolean>({
+          promise: (async (): Promise<boolean> => {
+            try {
+              return await collector.isReady();
+            } catch (err) {
+              this.logger.debug(
+                `Collector failed to get ready` +
+                  `${this.maximumWaitTimeForAllCollectorsInS}s and will return data from all collectors that are ready.`
+              );
+              return false;
+            }
+          })(),
+          timeoutMs: this.maximumWaitTimeForAllCollectorsInS * secondInMs,
+        });
+
+        return { isReadyWithTimeout, collector };
       })
     );
 
-    const collectorsTypesNotReady = collectorsWithStatus
-      .filter((collectorWithStatus) => collectorWithStatus.isReady === false)
-      .map((collectorWithStatus) => collectorWithStatus.collector.type);
-
-    const allReady = collectorsTypesNotReady.length === 0;
-
-    if (!allReady && this.maximumWaitTimeForAllCollectorsInS >= 0) {
-      const nowTimestamp = +new Date();
-      this._waitingForAllCollectorsTimestamp =
-        this._waitingForAllCollectorsTimestamp || nowTimestamp;
-      const timeWaitedInMS = nowTimestamp - this._waitingForAllCollectorsTimestamp;
-      const timeLeftInMS = this.maximumWaitTimeForAllCollectorsInS * 1000 - timeWaitedInMS;
-      if (timeLeftInMS <= 0) {
-        this.logger.debug(
-          `All collectors are not ready (waiting for ${collectorsTypesNotReady.join(',')}) ` +
-            `but we have waited the required ` +
-            `${this.maximumWaitTimeForAllCollectorsInS}s and will return data from all collectors that are ready.`
-        );
-
-        return true;
-      } else {
-        this.logger.debug(`All collectors are not ready. Waiting for ${timeLeftInMS}ms longer.`);
-      }
-    } else {
-      this._waitingForAllCollectorsTimestamp = undefined;
+    const hasTimedOutCollectors = collectorsWithStatus.some(
+      (collectorWithStatus) => collectorWithStatus.isReadyWithTimeout.timedout
+    );
+    if (hasTimedOutCollectors) {
+      this.logger.debug(
+        `Some collectors timedout getting ready. ` +
+          `Waited for ${this.maximumWaitTimeForAllCollectorsInS}s and will return data from collectors that are ready.`
+      );
```

Comment:
For troubleshooting, it'd be great to know which collectors timedout. Also, should we `log.warn` this?

Reply:
Added collector types to the log. I'd rather keep it on `debug` like most things in telemetry to avoid annoying and misleading users. Especially since this might log when the true issue is not the collector but failure to connect to ES or something.

Reply:
Fair enough! My concern is that devs don't usually run in `debug` unless they really need it. So we may not realise that the errors are happening. However, I agree that it shouldn't affect the user.

Let's tackle https://github.com/elastic/kibana/issues/89839, and we'll be able to move these to `warn`. What do you think?